content/blog/2023/2023-11-20-design-patterns.md

```diff
@@ -0,0 +1,641 @@
+interface ProductB {
+    fun operationB(): String
+}
+
+// Concrete Product B1
+class ConcreteProductB1 : ProductB {
+    override fun operationB(): String {
+        return "Product B1"
+    }
+}
+
+// Concrete Product B2
+class ConcreteProductB2 : ProductB {
+    override fun operationB(): String {
+        return "Product B2"
+    }
+}
+
+// Abstract Factory
+interface AbstractFactory {
+    fun createProductA(): ProductA
+    fun createProductB(): ProductB
+}
+
+// Concrete Factory 1
+class ConcreteFactory1 : AbstractFactory {
+    override fun createProductA(): ProductA {
+        return ConcreteProductA1()
+    }
+
+    override fun createProductB(): ProductB {
+        return ConcreteProductB1()
+    }
+}
+
+// Concrete Factory 2
+class ConcreteFactory2 : AbstractFactory {
+    override fun createProductA(): ProductA {
+        return ConcreteProductA2()
+    }
+
+    override fun createProductB(): ProductB {
+        return ConcreteProductB2()
+    }
+}
+
+// Client Code
+fun main() {
+    val factory1: AbstractFactory = ConcreteFactory1()
+    val productA1: ProductA = factory1.createProductA()
+    val productB1: ProductB = factory1.createProductB()
+
+    println(productA1.operationA()) // Output: Product A1
+    println(productB1.operationB()) // Output: Product B1
+
+    val factory2: AbstractFactory = ConcreteFactory2()
+    val productA2: ProductA = factory2.createProductA()
+    val productB2: ProductB = factory2.createProductB()
+
+    println(productA2.operationA()) // Output: Product A2
+    println(productB2.operationB()) // Output: Product B2
+}
+
+```
+
+In this example, `AbstractFactory` declares the creation methods for two types of products `ProductA` and `ProductB`. Concrete factories `ConcreteFactory1` and `ConcreteFactory2` implement these creation methods to produce specific products `ConcreteProductA1`, `ConcreteProductA2`, `ConcreteProductB1` and `ConcreteProductB2`. The client code can then use a specific factory to create products without needing to know the concrete classes of those products.
+
+This structure allows for easy extension of the system by introducing new products and factories without modifying the existing client code.
+
+## Factory Design Pattern
+
+The Factory Design Pattern is a creational pattern that provides an interface for creating objects in a super class but allows subclasses to alter the type of objects that will be created. This pattern is often used when a class cannot anticipate the class of objects it must create.
+
+Here's an example of a simple Factory Design Pattern in Kotlin:
+
+```kotlin
+// Product interface
+interface Product {
+    fun create(): String
+}
+
+// Concrete Product A
+class ConcreteProductA : Product {
+    override fun create(): String {
+        return "Product A"
+    }
+}
+
+// Concrete Product B
+class ConcreteProductB : Product {
+    override fun create(): String {
+        return "Product B"
+    }
+}
+
+// Factory interface
+interface ProductFactory {
```

Comment:
Where is the difference to the Abstract Factory above? Why have this interface here?